### DISCUSSION TITLE
How to implement a LDS server with Milo?

### DISCUSSION THREAD

**Discussion by ashhassan100** [eclipse-milo/milo#1015] [General]
Hi. I am completely new to OPC UA and have been reading the documentation provided by the OPC foundation. I am working on a project that requires an implementation of LDS and I had a few questions.

1) Is the LDS itself a OPC UA Sever?
2) When a OPC UA Server registers itself to the LDS, does it call the method registerServer() provided by the LDS? Does it pass the endpoint url of the LDS? e.g. opc.tcp/LDSMachineName/4840?
3) When a server registers to the LDS, do I store the information about the servers in the AddressSpace?
4) Would I only have to include the discovery services if I just want the LDS to let OPC UA Servers register to it and LDS stores the information and allow clients to use findServers() to get the information about the servers?
5) How would someone go about implementing a LDS with MILO SDK? In brief steps

**Comment by kevinherron** [comment]
> Is the LDS itself a OPC UA Sever?

Yes.

>When a OPC UA Server registers itself to the LDS, does it call the method registerServer() provided by the LDS? Does it pass the endpoint url of the LDS? e.g. opc.tcp/LDSMachineName/4840?

It should pass its own endpoint URL, so other clients know where to find it.

>When a server registers to the LDS, do I store the information about the servers in the AddressSpace?

You have to store it somewhere... not sure about an AddressSpace or not. It's an implementation detail.

>Would I only have to include the discovery services if I just want the LDS to let OPC UA Servers register to it and LDS stores the information and allow clients to use findServers() to get the information about the servers?

You may need some other basic services as well, like Session and Attribute services, but I'm not sure.

>How would someone go about implementing a LDS with MILO SDK? In brief steps

Read everything in part 12 that pertains to Local Discovery: https://reference.opcfoundation.org/GDS/docs/

**Reply by ashhassan100** [reply]
>It should pass its own endpoint URL, so other clients know where to find it.
>
Sorry I am a bit confused. What I understand is that the LDS server writes the function registerServer() and the OPC UA Servers call on that function to register themselves with the LDS. Then, the Client can call findServers() (which is written in the LDS implementation) to get a list of information about the servers which is registered to the LDS. You mentioned the server should pass its own endpoint URL when calling registerServer() but then what's the point of the client calling getEndpoints? Also if the server registers to the LDS how does it create the connection? The LDS itself has a connection URL and so I am wondering if that also gets passed onto the registerServer() function as a parameter.

**Reply by kevinherron** [reply]
>You mentioned the server should pass its own endpoint URL when calling registerServer() but then what's the point of the client calling getEndpoints? 

When a server registers with a LDS it registers its discovery URL. This is the URL clients will use to call GetEndpoints against that server.

>Also if the server registers to the LDS how does it create the connection? The LDS itself has a connection URL and so I am wondering if that also gets passed onto the registerServer() function as a parameter.

Servers that will be registered with a LDS must be configured with knowledge about where that LDS is (its URL).

**Reply by ashhassan100** [reply]
> Servers that will be registered with a LDS must be configured with knowledge about where that LDS is (its URL).
>
This means that server already knows the URL to the LDS for it to register to. The LDS does not expose the URL. It's already pre-defined?

**Reply by kevinherron** [reply]
Yes, it must know the URL somehow. The default would be something like `opc.tcp://hostname:4840/UADiscovery`, but you still need to know the hostname or IP address. The original intention for an LDS would be that it runs on localhost (hence the L in Local), but this was a concept copied from OPC Classic/DA and OPC Enum and it’s not how people run OPC UA servers in practice.

**Comment by ashhassan100** [comment]
Could you please tell me if this is correct? I am trying to implement RegisterServers() and FindServers().

For RegisterServers(), a UA Server calls on the method implemented by the LDS. In that call, it passes in the RegisteredServer data type which contains the gatewayServerUri, and discoveryUrls []. The gatewayServerUri is the pre-defined URL of the LDS and the discoveryUrls [] is the DiscoveryEndpoints for the server (which is what the UA Client gets from the LDS to connect to the UA Server). Then, the LDS registers the UA Server and stores it in the LDS Server. Then, a client calls the FindServers() method which is implemented by the LDS. When the client makes the FindServers() call, it passes in the endpointUrl which is again the same as the gatewayServerUri that it contains the URL to connect to the LDS. Then the FindServers() returns the data type ApplicationDescription which contains discoveryUrls [] which allows the client to then call the GetEndPoints using that discoverURL on the actual server and it returns EndpointDescription which contains endpointUrl and that is the FINAL endpoint needed by the client to connect to the server.

Is this correct? I read part 12 of the reference and I am very confused on the different types of URLs, discovery endpoint, session endpoint, etc... What I am understanding is that every server has a discovery endpoint (to obtain information on how to connect. in the case of an LDS, client connects to discovery endpoint of LDS to obtain discoveryURL of UA Server which in turn the client uses that obtained discoveryURL to call getendpoints and obtain the actual URL to connect to it) and an endpoint to actually connect to it.

Also, I was thinking of implementing LDS and there's an issue of running the discovery service registerServer and findServer in the background. Do you have any suggestions to implement this? For example, eventually when I implement LDS, it will be run as a jar file and I would want a UA Server to register itself to the LDS. How do I make it so that the UA server can do that while the LDS is running as a jar file?

This was long and apologies for that. I would really appreciate if you give suggestions/ideas on this!
